ledger/participant-integration-api/src/main/scala/platform/apiserver/services/tracking/TrackerMap.scala

```diff
@@ -125,53 +140,56 @@ private[services] object TrackerMap {
-  final case object Closed extends AsyncResourceState[Nothing]
-  final case class Ready[T <: AutoCloseable](t: T) extends AsyncResourceState[T]
+  private sealed trait AsyncResourceState[+T <: AutoCloseable]
+  private final case class Waiting[T <: AutoCloseable](future: Future[T])
+      extends AsyncResourceState[T]
+  private final case class Ready[T <: AutoCloseable](resource: T) extends AsyncResourceState[T]
+  private final case object Closed extends AsyncResourceState[Nothing]
+  private final case class Failed(exception: Throwable) extends AsyncResourceState[Nothing]
 
   /** A holder for an AutoCloseable that can be opened and closed async.
     * If closed before the underlying Future completes, will close the resource on completion.
     */
   final class AsyncResource[T <: AutoCloseable](
-      future: Future[T]
+      start: Future[T]
   )(implicit val loggingContext: LoggingContext) {
-    private val logger = ContextualizedLogger.get(this.getClass)
-
-    // Must progress Waiting => Ready => Closed or Waiting => Closed.
-    val state: AtomicReference[AsyncResourceState[T]] = new AtomicReference(Waiting)
-
-    future.andThen {
-      case Success(t) =>
-        if (!state.compareAndSet(Waiting, Ready(t))) {
-          // This is the punch line of AsyncResource.
-          // If we've been closed in the meantime, we must close the underlying resource also.
-          // This "on-failure-to-complete" behavior is not present in scala or java Futures.
-          t.close()
-        }
-      // Someone should be listening to this failure downstream
-      // TODO(mthvedt): Refactor so at least one downstream listener is always present,
-      // and exceptions are never dropped.
-      case Failure(ex) =>
-        logger.error("failure to get async resource", ex)
-        state.set(Closed)
-    }(DirectExecutionContext)
-
-    def flatMap[U](f: T => Future[U])(implicit ex: ExecutionContext): Future[U] =
-      state.get() match {
-        case Waiting => future.flatMap(f)
-        case Closed => throw new IllegalStateException()
-        case Ready(t) => f(t)
-      }
+    // Must progress Waiting => Ready => Closed, Waiting => Closed, or Waiting => Failed.
+    private val state: AtomicReference[AsyncResourceState[T]] = new AtomicReference(
+      Waiting(
+        start.andThen {
+          case Success(resource) =>
+            state.set(Ready(resource))
+          case Failure(exception) =>
+            state.set(Failed(exception))
+        }(DirectExecutionContext)
+      )
+    )
 
-    def ifPresent[U](f: T => U): Option[U] = state.get() match {
-      case Ready(t) => Some(f(t))
-      case _ => None
-    }
+    private[TrackerMap] def currentState: AsyncResourceState[T] = state.get()
+
+    def withResource[U](f: T => Future[U])(implicit ex: ExecutionContext): Future[U] =
+      currentState match {
+        case Waiting(future) => future.flatMap(_ => withResource(f)) // try again
```

Comment:
Isn't this going to go in a loop like crazy? What's the benefit vs the previous behavior?

Reply:
Nope, it will wait on the future and then run again, which guarantees that the state will never be `Waiting` again.

The previous behavior would just sometimes do nothing, which was… not OK.

Reply:
I still fail to see the difference. I assume this is a replacement for the previous `flatMap` implementation, where we would flatMap the resulting future directly, now we flatMap it through an intermediary call to `withResource`. If the `future` never completes it will still do nothing.

Reply:
If the future never completes, it will do nothing. The previous behavior would just discard `f` entirely if it was in a waiting state. The new behavior waits on `future`, which guarantees that `state` will become `Ready` or `Failed`, and _then_ tries again.

Reply:
Ok, I need to take a closer look at the state transitions and how does the `future` changes between them as I had other conceptions about it. 

No reason to block this PR because of this though.